localstack/services/cloudformation/models/ec2.py

```diff
@@ -147,9 +147,16 @@ def fetch_state(self, stack_name, resources):
         gw_id = self.resolve_refs_recursively(stack_name, props.get("GatewayId"), resources)
         route_tables = client.describe_route_tables()["RouteTables"]
         route_table = ([t for t in route_tables if t["RouteTableId"] == table_id] or [None])[0]
+        subnet_id = self.resolve_refs_recursively(stack_name, props.get("SubnetId"), resources)
         if route_table:
             associations = route_table.get("Associations", [])
             association = [a for a in associations if a.get("GatewayId") == gw_id]
+            if subnet_id:
+                association = [
+                    a
+                    for a in associations
+                    if a.get("GatewayId") == gw_id and a.get("SubnetId") == subnet_id
```

Comment:
```suggestion
                    if a.get("SubnetId") == subnet_id
```

the gateway check is redundant here since we've already filtered by gateway id in line 153